### DISCUSSION TITLE
Include Resources from modules that inherit from ModuleControlBase

### DISCUSSION THREAD

**Discussion by mdmontesinos** [oqtane/oqtane.framework#4037] [Q&A]
I want to create a reusable component that encapsulates its required dependencies so that they are only loaded when the component first appears instead of in the theme.

This component shouldn't be displayed in the module list of the page builder, but rather be included inside other modules like a regular blazor component, so I inherited from ModuleControlBase (for the OqtaneIgnore). 

I saw that RichTextEdit does include its own dependencies by overriding the Resources property and I did the same, but the resources are not loaded nor included in the head/body. The component itself is being correctly included in other modules, but without any js or css.

EDIT: If the resources are included in the parent module (a regular one with ModuleInfo) they are loaded, but as the component should be reusable among many modules, it should include its own dependencies.

Here's my "component":

```razor
@namespace MyNamespace
@inherits ModuleControlBase

<lite-youtube videoid="@VideoId" videotitle="@VideoTitle">
    <a class="lite-youtube-fallback" href=@($"https://www.youtube.com/watch?v={VideoId}")>Watch on YouTube - @VideoTitle</a>
</lite-youtube>

@code {
    [Parameter]
    public required string VideoId { get; set; }

    [Parameter]
    public string? VideoTitle { get; set; }

    public override string RenderMode => RenderModes.Static;

    public override List<Resource> Resources => new List<Resource>()
    {
        new Resource {
            ResourceType = ResourceType.Stylesheet,
            Url = "css/lite-youtube.css"
        },
        new Resource { 
            ResourceType = ResourceType.Script, 
            ES6Module = true, 
            Url = "https://cdn.jsdelivr.net/npm/@justinribeiro/lite-youtube@1.5.0/lite-youtube.js",
            CrossOrigin = "anonymous",
            Location = ResourceLocation.Body
        }
    };
}
```

**Comment by sbwalker** [comment]
I have explored this previously and it is much more difficult than it appears. For now you will need to declare your dependencies in the calling component - which I know is not optimal but there are much higher priorities at the moment.

**Reply by mdmontesinos** [reply]
But why does it work for RichTextEdit when included in the HtmlText module?

**Reply by sbwalker** [reply]
It works for JavaScript resources but not Stylesheet resources. This is because they are handled completely differently by the framework. The Html/Text Edit component defines its own Stylesheets related to the RichTextEditor.

**Reply by mdmontesinos** [reply]
JavaScript resources are also not working for me.

**Reply by sbwalker** [reply]
Log an issue. This is very likely related to Static rendering. However it may not get in the 5.1.0 release. For now take the approach that a Module needs to explicitly define its dependencies - which includes Resources. So if you are using a shared component which requires Resources then you will need to declare them in the module

**Reply by mdmontesinos** [reply]
Will do.

As a workaround for the time being, I've defined a static collection of the resources required by the reusable component, so they can be easily included in any module that uses it with the spread operator. In this way, you can also add "depencies" of many other child components with having to duplicate or hard-code the values.

Child component:

```razor
@namespace YourNamespace
@inherits ModuleControlBase

<lite-youtube videoid="@VideoId" videotitle="@VideoTitle">
    <a class="lite-youtube-fallback" href=@($"https://www.youtube.com/watch?v={VideoId}")>Watch on YouTube - @VideoTitle</a>
</lite-youtube>

@code {
    [Parameter]
    public required string VideoId { get; set; }

    [Parameter]
    public string? VideoTitle { get; set; }

    public override string RenderMode => RenderModes.Static;

    //Resources must be included in parent module (https://github.com/oqtane/oqtane.framework/discussions/4037)
    public static List<Resource> Dependencies => new List<Resource>()
    {
        new Resource {
            ResourceType = ResourceType.Stylesheet,
            Url = "css/lite-youtube.css",
        },
        //Obtained from: https://github.com/justinribeiro/lite-youtube
        new Resource { 
            ResourceType = ResourceType.Script, 
            ES6Module = true, 
            Url = "https://cdn.jsdelivr.net/npm/@justinribeiro/lite-youtube@1.5.0/lite-youtube.js",
            CrossOrigin = "anonymous",
            Location = ResourceLocation.Body,
            Reload = true
        }
    };
}
```

Parent ModuleInfo:
```c#
public class ModuleInfo : IModule
{
    public ModuleDefinition ModuleDefinition => new ModuleDefinition
    {
        //Module definition
        Resources =
        [
            new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/Module.css" },
            new Resource { ResourceType = ResourceType.Script, Url = "~/Module.js", Location = ResourceLocation.Body },
            .. ChildComponent.Dependencies,
        ]
    };
}
```

**Reply by mdmontesinos** [reply]
I've already logged the issue in #4038.

Also, I've also reported an issue concerning js resources for all modules in static mode.
